<a href="https://colab.research.google.com/github/amandarice97/data-analytics/blob/main/danl-m2-hw1-Rice-Amanda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Import the pandas Library
*import pandas as pd*

Below is nhl DataFrame that reads the file NHL-1617.csv containing data of all non-goalie NHL players who had at least some ice time, i.e. playing time, in the 2016/2017 regular season - without play-offs.

nhl = pd.read_csv('https://bcdanl.github.io/data/NHL-1617.csv')

#Variable Description
- id_player: Player ID
- Born: Date of Birth
- City: City of Birth
- Cntry: Country of Birth
- Nat: Nationality
- Last_Name: Last name
- First_name: First Name
- Position: Position
- Team: Team
- GP: The number of games
- G: The number of goals
- A: The number of assists
- TOI: The total time on ice (in second)
- TOI_GP: Average amount of playing time per game (in minute)

#Homework 2
*Pandas Basics - Loading, Summarizing, Selecting, Counting, and Sorting Data*

- A CSV (comma-separated values) is a plain-text file that uses a comma to separate values (e.g., nba.csv).
- The CSV is widely used for storing data, and we will use this throughout the module.
- We use the read_csv() function to load a CSV data file.
- The DataFrame is the workhorse of the pandas library and the data structure.

In [1]:
import pandas as pd
nhl = pd.read_csv('https://bcdanl.github.io/data/NHL-1617.csv')
type(nhl)
nhl

id_player        Born            City Cntry  Nat  Ht   Wt    Last_Name  \
0            1  30.04.1988        Hamilton   CAN  CAN  69  170       Abbott   
1            2  25.02.1987        Muskegon   USA  USA  74  218   Abdelkader   
2            3  23.09.1993       Stockholm   SWE  SWE  71  196        Aberg   
3            4  01.12.1991        Johnston   USA  USA  70  208      Acciari   
4            5  30.04.1992      Morristown   USA  USA  72  202     Agostino   
..         ...         ...             ...   ...  ...  ..  ...          ...   
883        884  18.04.1993        Huddinge   SWE  SWE  74  215    Zibanejad   
884        885  01.09.1987         Toronto   CAN  CAN  71  180  Zolnierczyk   
885        886  01.09.1987            Oslo   NOR  NOR  67  179   Zuccarello   
886        887  16.01.1992   Newport Beach   USA  USA  71  187       Zucker   
887        888  15.05.1995  St. Petersburg   RUS  RUS  73  224        Zykov   

    First_Name Position Team  GP   G   A    TOI  TOI_GP  
0      Spencer       LW  CHI   1   0   0    514    8.57  
1       Justin    LW/RW  DET  64   7  14  63969   16.65  
2       Pontus       LW  NSH  15   1   1  11102   12.33  
3         Noel        C  BOS  29   2   3  18047   10.23  
4        Kenny       LW  STL   7   1   2   5366   12.78  
..         ...      ...  ...  ..  ..  ..    ...     ...  
883       Mika     C/RW  NYR  56  14  23  57362   17.07  
884      Harry       LW  NSH  24   2   2  12776    8.87  
885       Mats  RW/C/LW  NYR  80  15  44  90378   18.83  
886      Jason    LW/RW  MIN  79  22  25  72455   15.28  
887   Valentin       LW  CAR   2   1   0    750    6.25  

[888 rows x 16 columns]

#Question 1
How many players had at least some ice time?

*Answer*

888 players had at least some ice time, measured in seconds. Player Ryan Pulock had the least amount of ice time at 237 seconds. Player Drew Doughty had the most amount of ice time at 133,550 seconds.

**Code Used to Answer Question 1:**
- nhl.describe() to generate descriptive statistics that summarize the central tendency, dispersion, and distribution of each variable.
- nhl.describe(include='all') to process string-type variables.
- nhl.sort_values(["TOI"], ascending=True) to sort ice time (seconds) by ascending order.
- nhl.sort_values(["TOI"], ascending=False) to sort ice time (seconds) by descending order.
- nhl.nsmallest(5, 'TOI') and nhl.nlargest(5, 'TOI') to identify the 5 smallest and largest values for ice time.

In [9]:
nhl.describe() #min = 237.00

id_player          Ht          Wt          GP           G           A  \
count  888.000000  888.000000  888.000000  888.000000  888.000000  888.000000   
mean   444.500000   73.082207  200.793919   49.858108    7.549550   12.656532   
std    256.487816    2.102503   15.107405   29.512883    8.608602   12.760311   
min      1.000000   66.000000  157.000000    1.000000    0.000000    0.000000   
25%    222.750000   72.000000  190.000000   19.000000    1.000000    2.000000   
50%    444.500000   73.000000  200.000000   60.000000    4.000000    9.000000   
75%    666.250000   75.000000  210.250000   78.000000   12.000000   19.000000   
max    888.000000   81.000000  265.000000   82.000000   44.000000   70.000000   

                 TOI      TOI_GP  
count     888.000000  888.000000  
mean    49343.443694   14.987297  
std     35036.726593    4.265842  
min       237.000000    3.950000  
25%     14740.250000   11.727500  
50%     51361.000000   14.880000  
75%     79144.750000   18.030000  
max    133550.000000   27.450000

In [11]:
nhl.describe(include='all') #888.00 players have had at least some ice time

/usr/local/lib/python3.10/dist-packages/lida/components/summarizer.py:74: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  cast_date_col = pd.to_datetime(df[column], errors='coerce')


id_player        Born     City Cntry  Nat          Ht          Wt  \
count   888.000000         888      888   888  888  888.000000  888.000000   
unique         NaN         797      497    19   17         NaN         NaN   
top            NaN  22.02.1991  Toronto   CAN  CAN         NaN         NaN   
freq           NaN           3       36   415  413         NaN         NaN   
mean    444.500000         NaN      NaN   NaN  NaN   73.082207  200.793919   
std     256.487816         NaN      NaN   NaN  NaN    2.102503   15.107405   
min       1.000000         NaN      NaN   NaN  NaN   66.000000  157.000000   
25%     222.750000         NaN      NaN   NaN  NaN   72.000000  190.000000   
50%     444.500000         NaN      NaN   NaN  NaN   73.000000  200.000000   
75%     666.250000         NaN      NaN   NaN  NaN   75.000000  210.250000   
max     888.000000         NaN      NaN   NaN  NaN   81.000000  265.000000   

       Last_Name First_Name Position Team          GP           G           A  \
count        888        888      888  888  888.000000  888.000000  888.000000   
unique       812        388       20   78         NaN         NaN         NaN   
top        Smith       Ryan        D  BUF         NaN         NaN         NaN   
freq           8         22      296   34         NaN         NaN         NaN   
mean         NaN        NaN      NaN  NaN   49.858108    7.549550   12.656532   
std          NaN        NaN      NaN  NaN   29.512883    8.608602   12.760311   
min          NaN        NaN      NaN  NaN    1.000000    0.000000    0.000000   
25%          NaN        NaN      NaN  NaN   19.000000    1.000000    2.000000   
50%          NaN        NaN      NaN  NaN   60.000000    4.000000    9.000000   
75%          NaN        NaN      NaN  NaN   78.000000   12.000000   19.000000   
max          NaN        NaN      NaN  NaN   82.000000   44.000000   70.000000   

                  TOI      TOI_GP  
count      888.000000  888.000000  
unique            NaN         NaN  
top               NaN         NaN  
freq              NaN         NaN  
mean     49343.443694   14.987297  
std      35036.726593    4.265842  
min        237.000000    3.950000  
25%      14740.250000   11.727500  
50%      51361.000000   14.880000  
75%      79144.750000   18.030000  
max     133550.000000   27.450000

In [14]:
nhl.sort_values(["TOI"], ascending=True) #sort values, min = 237

id_player        Born      City Cntry  Nat  Ht   Wt    Last_Name  \
650        651  06.10.1994   Dauphin   CAN  CAN  74  215       Pulock   
460        461  16.02.1989   Toronto   CAN  CAN  70  203      Liambas   
671        672  25.03.1992    Zagreb   HRV  HRV  74  200     Rendulic   
335        336  25.07.1993  Val-d'Or   CAN  CAN  76  210       Henley   
584        585  29.07.1994   Halifax   CAN  CAN  73  215      O'Brien   
..         ...         ...       ...   ...  ...  ..  ...          ...   
392        393  31.05.1990  Landsbro   SWE  SWE  72  191     Karlsson   
678        679  27.10.1994     Turku   FIN  FIN  76  203  Ristolainen   
94          95  27.03.1985    Roseau   USA  USA  77  260    Byfuglien   
797        798  21.01.1985   Madison   USA  USA  74  206        Suter   
181        182  08.12.1989    London   CAN  CAN  73  195      Doughty   

    First_Name Position Team  GP   G   A     TOI  TOI_GP  
650       Ryan        D  NYI   1   0   0     237    3.95  
460       Mike       LW  NSH   1   0   0     282    4.70  
671      Borna     RW/C  VAN   1   0   0     291    4.85  
335     Samuel     C/RW  COL   1   1   0     318    5.30  
584       Liam     C/LW  WSH   1   0   0     368    6.13  
..         ...      ...  ...  ..  ..  ..     ...     ...  
392       Erik        D  OTT  77  17  54  123989   26.83  
678     Rasmus        D  BUF  79   6  39  125458   26.47  
94      Dustin     D/RW  WPG  80  13  39  131734   27.45  
797       Ryan        D  MIN  82   9  31  132469   26.93  
181       Drew        D  L.A  82  12  32  133550   27.15  

[888 rows x 16 columns]

In [15]:
nhl.sort_values(["TOI"], ascending=False) #sort values, max = 133550

id_player        Born      City Cntry  Nat  Ht   Wt    Last_Name  \
181        182  08.12.1989    London   CAN  CAN  73  195      Doughty   
797        798  21.01.1985   Madison   USA  USA  74  206        Suter   
94          95  27.03.1985    Roseau   USA  USA  77  260    Byfuglien   
678        679  27.10.1994     Turku   FIN  FIN  76  203  Ristolainen   
392        393  31.05.1990  Landsbro   SWE  SWE  72  191     Karlsson   
..         ...         ...       ...   ...  ...  ..  ...          ...   
584        585  29.07.1994   Halifax   CAN  CAN  73  215      O'Brien   
335        336  25.07.1993  Val-d'Or   CAN  CAN  76  210       Henley   
671        672  25.03.1992    Zagreb   HRV  HRV  74  200     Rendulic   
460        461  16.02.1989   Toronto   CAN  CAN  70  203      Liambas   
650        651  06.10.1994   Dauphin   CAN  CAN  74  215       Pulock   

    First_Name Position Team  GP   G   A     TOI  TOI_GP  
181       Drew        D  L.A  82  12  32  133550   27.15  
797       Ryan        D  MIN  82   9  31  132469   26.93  
94      Dustin     D/RW  WPG  80  13  39  131734   27.45  
678     Rasmus        D  BUF  79   6  39  125458   26.47  
392       Erik        D  OTT  77  17  54  123989   26.83  
..         ...      ...  ...  ..  ..  ..     ...     ...  
584       Liam     C/LW  WSH   1   0   0     368    6.13  
335     Samuel     C/RW  COL   1   1   0     318    5.30  
671      Borna     RW/C  VAN   1   0   0     291    4.85  
460       Mike       LW  NSH   1   0   0     282    4.70  
650       Ryan        D  NYI   1   0   0     237    3.95  

[888 rows x 16 columns]

In [27]:
nhl.nsmallest(5, 'TOI') #5 smallest amount of ice time

id_player        Born      City Cntry  Nat  Ht   Wt Last_Name First_Name  \
650        651  06.10.1994   Dauphin   CAN  CAN  74  215    Pulock       Ryan   
460        461  16.02.1989   Toronto   CAN  CAN  70  203   Liambas       Mike   
671        672  25.03.1992    Zagreb   HRV  HRV  74  200  Rendulic      Borna   
335        336  25.07.1993  Val-d'Or   CAN  CAN  76  210    Henley     Samuel   
584        585  29.07.1994   Halifax   CAN  CAN  73  215   O'Brien       Liam   

    Position Team  GP  G  A  TOI  TOI_GP  
650        D  NYI   1  0  0  237    3.95  
460       LW  NSH   1  0  0  282    4.70  
671     RW/C  VAN   1  0  0  291    4.85  
335     C/RW  COL   1  1  0  318    5.30  
584     C/LW  WSH   1  0  0  368    6.13

In [28]:
nhl.nlargest(5, 'TOI') #5 largest amount of ice time

id_player        Born      City Cntry  Nat  Ht   Wt    Last_Name  \
181        182  08.12.1989    London   CAN  CAN  73  195      Doughty   
797        798  21.01.1985   Madison   USA  USA  74  206        Suter   
94          95  27.03.1985    Roseau   USA  USA  77  260    Byfuglien   
678        679  27.10.1994     Turku   FIN  FIN  76  203  Ristolainen   
392        393  31.05.1990  Landsbro   SWE  SWE  72  191     Karlsson   

    First_Name Position Team  GP   G   A     TOI  TOI_GP  
181       Drew        D  L.A  82  12  32  133550   27.15  
797       Ryan        D  MIN  82   9  31  132469   26.93  
94      Dustin     D/RW  WPG  80  13  39  131734   27.45  
678     Rasmus        D  BUF  79   6  39  125458   26.47  
392       Erik        D  OTT  77  17  54  123989   26.83

#Question 2
Who is the top scorer in terms of goals?

*Answer*

Sydney Crosby is the top scorer in terms of goals (see below).

**Code Used to Answer Question 2:**
- nhl.describe(include='all') to generate descriptive statistics that summarize the central tendency, dispersion, and distribution of each variable and to process string-type variables.
- nhl.nlargest(5, 'G') to identify the largest number of goals scored by player.
- nhl.sort_values(["G"], ascending=False) to sort by the number of goals scored by player in descending order.

In [16]:
nhl.describe(include='all') #max number of goals = 44

/usr/local/lib/python3.10/dist-packages/lida/components/summarizer.py:74: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  cast_date_col = pd.to_datetime(df[column], errors='coerce')


id_player        Born     City Cntry  Nat          Ht          Wt  \
count   888.000000         888      888   888  888  888.000000  888.000000   
unique         NaN         797      497    19   17         NaN         NaN   
top            NaN  22.02.1991  Toronto   CAN  CAN         NaN         NaN   
freq           NaN           3       36   415  413         NaN         NaN   
mean    444.500000         NaN      NaN   NaN  NaN   73.082207  200.793919   
std     256.487816         NaN      NaN   NaN  NaN    2.102503   15.107405   
min       1.000000         NaN      NaN   NaN  NaN   66.000000  157.000000   
25%     222.750000         NaN      NaN   NaN  NaN   72.000000  190.000000   
50%     444.500000         NaN      NaN   NaN  NaN   73.000000  200.000000   
75%     666.250000         NaN      NaN   NaN  NaN   75.000000  210.250000   
max     888.000000         NaN      NaN   NaN  NaN   81.000000  265.000000   

       Last_Name First_Name Position Team          GP           G           A  \
count        888        888      888  888  888.000000  888.000000  888.000000   
unique       812        388       20   78         NaN         NaN         NaN   
top        Smith       Ryan        D  BUF         NaN         NaN         NaN   
freq           8         22      296   34         NaN         NaN         NaN   
mean         NaN        NaN      NaN  NaN   49.858108    7.549550   12.656532   
std          NaN        NaN      NaN  NaN   29.512883    8.608602   12.760311   
min          NaN        NaN      NaN  NaN    1.000000    0.000000    0.000000   
25%          NaN        NaN      NaN  NaN   19.000000    1.000000    2.000000   
50%          NaN        NaN      NaN  NaN   60.000000    4.000000    9.000000   
75%          NaN        NaN      NaN  NaN   78.000000   12.000000   19.000000   
max          NaN        NaN      NaN  NaN   82.000000   44.000000   70.000000   

                  TOI      TOI_GP  
count      888.000000  888.000000  
unique            NaN         NaN  
top               NaN         NaN  
freq              NaN         NaN  
mean     49343.443694   14.987297  
std      35036.726593    4.265842  
min        237.000000    3.950000  
25%      14740.250000   11.727500  
50%      51361.000000   14.880000  
75%      79144.750000   18.030000  
max     133550.000000   27.450000

In [26]:
nhl.nlargest(5, 'G') #5 largest number of goals scored

id_player        Born          City Cntry  Nat  Ht   Wt  Last_Name  \
149        150  07.08.1987  Cole Harbour   CAN  CAN  71  200     Crosby   
426        427  17.06.1993        Maykop   RUS  RUS  71  178   Kucherov   
499        500  17.09.1997    Scottsdale   USA  USA  75  216   Matthews   
485        486  11.05.1988       Halifax   CAN  CAN  69  181   Marchand   
802        803  13.12.1991     Yaroslavl   RUS  RUS  72  219  Tarasenko   

    First_Name Position Team  GP   G   A    TOI  TOI_GP  
149     Sidney        C  PIT  75  44  45  89450   19.88  
426     Nikita       RW  T.B  74  40  45  86320   19.43  
499     Auston        C  TOR  82  40  29  86745   17.53  
485       Brad     LW/C  BOS  80  39  46  93295   19.35  
802   Vladimir       RW  STL  82  39  36  90872   18.47

In [17]:
nhl.sort_values(["G"], ascending=False) #sort values, max number of goals = 44

id_player        Born          City Cntry  Nat  Ht   Wt  Last_Name  \
149        150  07.08.1987  Cole Harbour   CAN  CAN  71  200     Crosby   
499        500  17.09.1997    Scottsdale   USA  USA  75  216   Matthews   
426        427  17.06.1993        Maykop   RUS  RUS  71  178   Kucherov   
485        486  11.05.1988       Halifax   CAN  CAN  69  181   Marchand   
802        803  13.12.1991     Yaroslavl   RUS  RUS  72  219  Tarasenko   
..         ...         ...           ...   ...  ...  ..  ...        ...   
725        726  28.09.1987          Rome   USA  USA  77  228    Sestito   
386        387  12.08.1996          Orsk   RUS  RUS  74  194    Kamenev   
377        378  16.08.1990      Montrose   CAN  CAN  69  176      Jones   
374        375  14.02.1992       Sudbury   CAN  CAN  70  180   Johnston   
0            1  30.04.1988      Hamilton   CAN  CAN  69  170     Abbott   

    First_Name Position Team  GP   G   A    TOI  TOI_GP  
149     Sidney        C  PIT  75  44  45  89450   19.88  
499     Auston        C  TOR  82  40  29  86745   17.53  
426     Nikita       RW  T.B  74  40  45  86320   19.43  
485       Brad     LW/C  BOS  80  39  46  93295   19.35  
802   Vladimir       RW  STL  82  39  36  90872   18.47  
..         ...      ...  ...  ..  ..  ..    ...     ...  
725        Tom       LW  PIT  13   0   2   4315    5.53  
386  Vladislav     LW/C  NSH   2   0   0   1207   10.07  
377     Connor     C/LW  NYI   4   0   0   2843   11.85  
374       Ryan        D  MTL   7   0   0   4145    9.87  
0      Spencer       LW  CHI   1   0   0    514    8.57  

[888 rows x 16 columns]

#Question 3
How can we count the number of NHL player for each country?

*Answer*

Counting with .value_counts() to count the number of NHL players for each country (see below).

Used the code: nhl['Cntry'].value_counts() to count the number of occurrences of each unique value in a Series/DataFrame.

In [19]:
nhl['Cntry'].value_counts() #counts the number of occurrences of each unique value in a Series/DataFrame

CAN    415
USA    242
SWE     77
RUS     39
CZE     31
FIN     30
CHE     12
SVK     10
DNK      9
DEU      6
AUT      5
FRA      4
NOR      2
LVA      1
EST      1
SVN      1
GBR      1
HRV      1
ITA      1
Name: Cntry, dtype: int64

#Question 4
How many unique countires are in nhl.csv?

*Answer*

Counting with .nunique()

There are 19 unique countires in nhl.csv (see below).

Used the code: nhl[['Cntry']].nunique() to count the number of unique values in each variable in a DataFrame.

In [21]:
nhl[['Cntry']].nunique() #counts the number of unique values in each variable in a DataFrame

Cntry    19
dtype: int64

#Question 5
Which three nationalities have the highest number of players?

*Answer*

Canada (413), United States (243), and Sweden (79) have the highest number of players (see below).

Used the code: nhl['Nat'].value_counts() to count the number of occurrences of each unique value in a Series/DataFrame.

If Nat was a quantitative variable, we could instead use the code: nhl.nlargest (3, 'Nat').

In [25]:
nhl['Nat'].value_counts() #counts the number of occurrences of each unique value in a Series/DataFrame

CAN    413
USA    243
SWE     79
RUS     38
FIN     32
CZE     32
CHE     13
SVK     11
DNK      9
DEU      5
FRA      4
AUT      3
NOR      2
LVA      1
SVN      1
GBR      1
HRV      1
Name: Nat, dtype: int64